README.md
# clj-djl-examples

examples with [clj-djl](https://github.com/kimim/clj-djl).

## Installation

Install [leiningen](https://leiningen.org)

## Usage

Train MNIST dataset:

    $ lein run -m clj-djl-examples.train

Test with the model:

    $ lein run -m clj-djl-examples.predict

## License

Copyright © 2020 kimim

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2020-09-15
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2020-09-15
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/clj-djl-examples/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/clj-djl-examples/compare/0.1.0...0.1.1


test/clj_djl_examples/core_test.clj
(ns clj-djl-examples.core-test
  (:require [clojure.test :refer :all]
            [clj-djl-examples.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/clj_djl_examples/core.clj
(ns clj-djl-examples.core
  (:gen-class)
  (:require [clj-djl.ndarray :as ndarray]
            [clj-djl.model :as model]
            [clj-djl.model-zoo :as zoo]
            [clj-djl.training :as train]
            [clj-djl.engine :as engine]
            [clj-djl.training.dataset :as ds])
  (:import [ai.djl.basicdataset Mnist]))

(defn train-mnist []
  (let [mnist   ;;dataset
        (-> (Mnist/builder)
            (ds/set-sampling 32 true)
            (ds/build)
            (ds/prepare (train/new-progress-bar)))
        model   ;;model from model zoo
        (-> (model/new-instance "mlp")
            (model/set-block (zoo/mlp (* 28 28) 10 (int-array 128 64))))
        config  ;;trainer configs
        (-> (train/softmax-cross-entropy-loss)
            (train/new-training-config)
            (train/add-evaluator (train/new-accuracy))
            (train/add-training-listeners (train/new-default-training-listeners)))
        trainer  ;;trainer
        (-> (train/new-trainer model config)
            (train/initialize [(ndarray/shape 1 (* 28 28))]))]
    (doseq [epoch (range 1)] ;;train for 2 epoch
      (doseq [batch (train/iterate-dataset trainer mnist)]
        (train/train-batch trainer batch)
        (train/step trainer)
        (train/close batch))
      (train/notify-listeners trainer (fn [listner] (.onEpoch listner trainer))))))

(defn -main
  "I don't do a whole lot ... yet."
  [& args]
  (train-mnist)
  (println "Training mnist done!"))


src/clj_djl_examples/predict.clj
(ns clj-djl-examples.predict
  (:require [clj-djl.modality.cv :as cv]
            [clj-djl.model :as model]
            [clj-djl.ndarray :as ndarray]
            [clj-djl.model-zoo :as zoo]))

(defn -main
  "I don't do a whole lot ... yet."
  [& args]
  (let [img (cv/download-image-from "https://djl-ai.s3.amazonaws.com/resources/images/0.png")
        model
        (-> (model/new-instance "mlp")
            (model/set-block (zoo/mlp (* 28 28) 10 (int-array 128 64)))
            (model/load "build/mlp"))
        translator
        (-> (ai.djl.modality.cv.translator.ImageClassificationTranslator/builder)
            (.addTransform (ai.djl.modality.cv.transform.ToTensor.))
            (.optSynset (->> (range 0 10)
                             (map str)))
            (.build))
        predicator (model/new-predictor
                    model
                    translator)]
    (println (str (model/predict predicator img)))))


src/clj_djl_examples/train.clj
(ns clj-djl-examples.train
  (:gen-class)
  (:require [clojure.java.io :as io]
            [clj-djl.ndarray :as ndarray]
            [clj-djl.model :as model]
            [clj-djl.model-zoo :as zoo]
            [clj-djl.training :as train]
            [clj-djl.engine :as engine]
            [clj-djl.training.dataset :as ds])
  (:import [ai.djl.basicdataset Mnist]
           [java.nio.file Paths]))

(defn train-mnist []
  (let [mnist   ;;dataset
        (-> (Mnist/builder)
            (ds/set-sampling 32 true)
            (ds/build)
            (ds/prepare (train/new-progress-bar)))
        model   ;;model from model zoo
        (-> (model/new-instance "mlp")
            (model/set-block (zoo/mlp (* 28 28) 10 (int-array 128 64))))
        config  ;;trainer configs
        (-> (train/softmax-cross-entropy-loss)
            (train/new-training-config)
            (train/add-evaluator (train/new-accuracy))
            (train/add-training-listeners (train/new-default-training-listeners)))
        trainer  ;;trainer
        (-> (train/new-trainer model config)
            (train/initialize [(ndarray/shape 1 (* 28 28))]))]
    (doseq [epoch (range 3)] ;;train for 3 epoch
      (doseq [batch (train/iterate-dataset trainer mnist)]
        (train/train-batch trainer batch)
        (train/step trainer)
        (train/close batch))
      (train/notify-listeners trainer (fn [listner] (.onEpoch listner trainer))))
    model))


(defn save-model [model path]
  (let [nio-path (java.nio.file.Paths/get path (into-array [""]))]
    (io/make-parents path)
    (model/set-property model "Epoch" "3")
    (model/save model nio-path "mlp")))


(defn -main
  "I don't do a whole lot ... yet."
  [& args]
  (-> (train-mnist)
      (save-model "build/mlp"))
  (println "model trained and saved for mnist dataset!"))
